Combine `Elsif` node with `IfNode`

egiurleo opened the pull request:
Related to #278. Completed with @bitwise-aiden.

This PR eliminates the `Elsif` node and folds its functionality into `IfNode`. Now, `elsif` conditional statements will be represented with `IfNodes` that are the consequent of another `IfNode`. This eliminates an unnecessary node type from syntax tree, simplifies a lot of logic, and brings syntax tree closer to other AST representations of Ruby.

egiurleo commented on lib/syntax_tree/parser.rb:
```diff
@@ -1491,11 +1491,12 @@ def on_elsif(predicate, statements, consequent)
         consequent: consequent,
-        location: beginning.location.to(ending.location)
+        location: beginning.location.to(ending.location),
+        beginning: beginning.value
```
I made a couple choices here:
- I named this argument `beginning` because it seemed that other nodes also had an attribute with the same name, and I wanted to follow existing patterns.
- I decided to pass the value of the keyword as a string because there are other places where `IfNode`s are constructed manually, and I wanted to make it easy to create a value to pass in here.

Let me know if those both sound okay!

kddnewton replied:
`beginning` is fine as a name, but I think I've used `keyword` in other places.

I see what you mean about making it easy to be constructed, but I think just to keep consistency I'd prefer passing the whole keyword.

egiurleo commented on lib/syntax_tree/node.rb:
```diff
@@ -6264,7 +6175,7 @@ def format(q)
         # print a special version that uses the ternary operator if it fits on
         # one line.
-        if Ternaryable.call(q, node)
+        if Ternaryable.call(q, node) && keyword != "elsif"
```
One of the test cases was turning an `elsif` into a ternary, so I added this, but happy to modify the `Ternatyable.call` method in some way to prevent this from happening instead if that makes more sense.

kddnewton replied:
I think it shouldn't even get into here honestly. I think you can just copy-paste what used to be in the elsif formatting and put it into the `If` node's formatting function, and then call into `ConditionalFormatter` if it's not an elsif. That makes it so you don't have to have all of these checks floating around.

egiurleo commented on lib/syntax_tree/node.rb:
```diff
@@ -6322,8 +6233,10 @@ def format_break(q, force:)
 
-      force ? q.breakable_force : q.breakable_space
-      q.text("end")
+      unless keyword == "elsif"
```
I added this conditional here because an `elsif` is always a consequent, so it should never be adding the `end` keyword to the end of the conditional block. Let me know if there's a better way to do this!

egiurleo commented on lib/syntax_tree/translation/parser.rb:
```diff
@@ -1381,7 +1351,8 @@ def visit_if(node)
-              when Elsif
+              when IfNode
+                # The `IfNode` represents an `elsif` clause
                 source_range_length(node.consequent.location.start_char, 5)
```
I think it's correct here to still use the next 5 characters for the `else_token` because the token will still be `elsif`.

egiurleo commented on lib/syntax_tree/translation/parser.rb:
```diff
@@ -238,7 +238,7 @@ def visit_aryptn(node)
             children << s(:match_rest, [visit(node.rest)], nil)
           elsif node.posts.empty? &&
-                node.rest.location.start_char == node.rest.location.end_char
+                  node.rest.location.start_char == node.rest.location.end_char
```
This change was made by syntax tree

egiurleo commented on lib/syntax_tree/translation/parser.rb:
```diff
@@ -409,7 +409,8 @@ def visit_begin(node)
           s(:kwbegin, [], location)
         elsif node.bodystmt.rescue_clause.nil? &&
-              node.bodystmt.ensure_clause.nil? && node.bodystmt.else_clause.nil?
+                node.bodystmt.ensure_clause.nil? &&
+                node.bodystmt.else_clause.nil?
```
So was this one

egiurleo commented on lib/syntax_tree/visitor/field_visitor.rb:
```diff
@@ -352,15 +352,6 @@ def visit_else(node)
         end
       end
 
-      def visit_elsif(node)
-        node(node, "elsif") do
-          field("predicate", node.predicate)
-          field("statements", node.statements)
-          field("consequent", node.consequent) if node.consequent
-          comments(node)
-        end
-      end
-
```
I don't think this file or any of its subclasses have tests, so not 100% confident in this change. It looks like the logic in the `visit_if` method is almost identical, but I might be missing some nuance here.

egiurleo commented on lib/syntax_tree/visitor/mutation_visitor.rb:
```diff
@@ -322,14 +322,6 @@ def visit_else(node)
         )
       end
 
-      # Visit a Elsif node.
-      def visit_elsif(node)
-        node.copy(
-          statements: visit(node.statements),
-          consequent: visit(node.consequent)
-        )
-      end
-
```
This one also doesn't seem to have any tests. The only difference between this method and `visit_if` is that `visit_if` copies the `predicate`:

```
      def visit_if(node)
        node.copy(
          predicate: visit(node.predicate),
          statements: visit(node.statements),
          consequent: visit(node.consequent)
        )
      end
```

It seems like that should also be done for `elsif`, but maybe it shouldn't be for some reason?

kddnewton replied:
Nah it should, I think it was missed when I templated out that method.

egiurleo commented:
Hey @kddnewton just gently bumping this. I'd love a review when you have time!

kddnewton commented on lib/syntax_tree/node.rb:
```diff
@@ -6420,17 +6333,21 @@ class IfNode < Node
 
-    def initialize(predicate:, statements:, consequent:, location:)
+    # [ String ] the opening of the conditional statement
+    attr_reader :beginning
```
Instead of attaching the value of the token, could you attach the token itself? That's consistent with some other places where we've done this. Also, would you be able to make it the first child mentioned/the first one in `initialize` and `copy`? That's just closer to YARP so it's the directly we're going anyway.

kddnewton commented on lib/syntax_tree/node.rb:
```diff
@@ -6463,17 +6387,19 @@ def deconstruct_keys(_keys)
 
     def format(q)
-      ConditionalFormatter.new("if", self).format(q)
+      ConditionalFormatter.new(beginning, self).format(q)
```
I mentioned this elsewhere, but I think we should check if it's an elsif here and do that formatting here, otherwise call into `ConditionalFormatter`

kddnewton reviewed (changes requested):
Just a couple of changes, thanks so much for this work!

egiurleo commented:
@kddnewton Thanks for the feedback! I think I've implemented all the changes you requested. The big ones were:
* I changed the `beginning` attr on `IfNode` to be named `keyword` and kept it as a keyword rather than a string
* I changed the formatting to just copy-paste the old `Elsif` formatting if the node represents an elsif

kddnewton reviewed (approved):
This is so good!

kddnewton commented:
@egiurleo I ended up merging this into the `structure` branch because it's a breaking change and I'd like to make more breaking changes, but in case there are bugs off `main` I'd still like to be able to release. So I'm going to keep `structure` around until we decide to release the breaking changes all at once.
